### ISSUE TITLE
VisualStudio (17.4. Preview 5) fails on language features for projects referencing Fsharp.Compiler.Services

### ISSUE THREAD

**Issue by T-Gro** [dotnet/fsharp#14179]
When opening the VisualFsharp.sln, build works just fine.
Language services within the Fsharp.Compiler.Services work just fine as well.

However, when going into a project that directly or indirectly depends on Fsharp.Compiler.Service, language features stopped to work, demonstrated by a Object reference exception coming out of it.

When debugging the issue, the first failure happens on a pattern match of an Entity that that is unlinked (a pattern match, even though it has a | _ -> , tries to access .get_Tag and crashes on that).
Note that even when I tried to defensively cover the cases on Entity properties, the error just happens a little later or something else. The issue is more likely in the existance of such unlinked entity itself.

Test case:
- Open VisualFsharp.sln\
-  Go to any test file in component tests project, e.g. "EmptyArray.fs"
- Any references to Test.Utilities are failing
- Within the Test.Utilities project, the first error (coloring stops) is at FileSystemShim usage - but that just happens to be the first usage of any type in Test.Utilities.

<img width="1774" alt="image" src="https://user-images.githubusercontent.com/46543583/198010418-65aaaf81-e3b4-4a2f-9d8a-3eb67848b86d.png">

**Comment by vzarytovskii** [comment]
One thing to look at - https://github.com/dotnet/fsharp/pull/13671

Since it's related to pickling

**Comment by T-Gro** [comment]
This does not happen at classical build.
The code path being hit when loading a project references that results into thousands of Unlinked entities is via Incremental build:

```
type RawFSharpAssemblyDataBackedByLanguageService (tcConfig, tcGlobals, generatedCcu: CcuThunk, outfile, topAttrs, assemblyName, ilAssemRef) =

    let exportRemapping = MakeExportRemapping generatedCcu generatedCcu.Contents

    let sigData =
        let _sigDataAttributes, sigDataResources = EncodeSignatureData(tcConfig, tcGlobals, exportRemapping, generatedCcu, outfile, true)
        [ for r in sigDataResources  do
            GetResourceNameAndSignatureDataFunc r
        ]

```

**Comment by dsyme** [comment]
If the error is indeed related to metadata compression, some thoughts are

* Try to remove metadata compression in-repo, so if it repros
* Remove it selectively for FCS, see if it repros. 
* Work out if the VS it repros in supports compressed metadata at all. That is, `.\build /deploy` RoslynDev should, others will depend on release cycle

**Comment by T-Gro** [comment]
This fails for me on:
+17.4. Preview 5
+Latest dogfood VS
+as well as when launching VisualFsharpDebug on latest main, i.e. RoslynDev


I have tried removing compression everywhere, and it still reproduces => compression was a false suspect.

When inspecting the code path, metadata was not getting compressed.
The call to EncodeSignatureData really produced an uncompressed byte array (a big one), but for some reason GetResourceNameAndSignatureDataFunc  cannot cope with it just a few milliseconds later.

**Comment by T-Gro** [comment]
And I still did not spot what is making the behaviour of RawFSharpAssemblyDataBackedByLanguageService  so different from regular build (because regular build works just fine).

And, why it fails only for Fsharp.Compiler.Service when used as a reference, but not for other projects.
Size? Any capacity-limited buffers ? Overflow of some auto-increment byte/uint/.. perharps?

**Comment by T-Gro** [comment]
@dsyme :
Would you mind looking at this together, or suggesting other tips on how to spot issues around picking+unpickling?
Please let me know which way to connect is preferred by you.

**Comment by dsyme** [comment]
@T-Gro I have time now if you do, else put something on my calendar for tomorrow?

**Comment by dsyme** [comment]
Looking around, using

    .\build -c Debug -deploy
    devenv FSharp.sln /RootSuffix RoslynDev

I repro the failure via the assert that pops up.  The FSharp.Compiler.Service metadata seems to be populated in `tcImports`

![image](https://user-images.githubusercontent.com/7204669/198382126-4477baa8-c622-4c35-ba43-f9b43f898f75.png)

So agreed compression or the basics of metadata unpickling doesn't seem to be the problem.  Will look at the specific assert now

**Comment by dsyme** [comment]
Looking at this frame of the assert stack:

![image](https://user-images.githubusercontent.com/7204669/198385798-bd52e7a5-30b8-460b-b9a3-f60cec29e3fd.png)

the metadata for "pos" looks fine, and this is not a type abbreviationnor IsMeasurableReprTycon

![image](https://user-images.githubusercontent.com/7204669/198386287-29217843-abc3-4243-8c51-0a672a88a586.png)

![image](https://user-images.githubusercontent.com/7204669/198386425-99272991-c9ca-4527-9d3f-de4d22646e18.png)

Very strange as the `stripTyEqnsA` should just return

**Comment by dsyme** [comment]
And the TypeReprInfo is fine:

![image](https://user-images.githubusercontent.com/7204669/198387962-0aac9fa1-3e91-47b5-ac97-bc796ec416a9.png)

**Comment by vzarytovskii** [comment]
@dsyme 
@KevinRansom figured that the following PR caused the issue:
https://github.com/dotnet/fsharp/commit/51635bbebf36e39a86d06d96730e13cfc40a095f

**Comment by vzarytovskii** [comment]
@dsyme 
@KevinRansom bisected that the following PR caused the issue:
https://github.com/dotnet/fsharp/commit/51635bbebf36e39a86d06d96730e13cfc40a095f

**Comment by dsyme** [comment]
Well well well, I wonder why

**Comment by vzarytovskii** [comment]
I'm still trying to figure out why ccus are affected only in our incremental builder scenarios in FCS, and not in standalone compiler.

**Comment by dsyme** [comment]
I can't see anything problematic in that PR as yet.

**Comment by dsyme** [comment]
Ok, so I see, pos.TypeAbbrev is set to a Tycon that has not been fixed up. I'll try to work out why

![image](https://user-images.githubusercontent.com/7204669/198393493-125611c3-62cc-4eb9-bc5e-b55ae2627fee.png)

**Comment by dsyme** [comment]
> @KevinRansom bisected that the following commit caused the issue:

We're 100% certain of this?

**Comment by vzarytovskii** [comment]
> > @KevinRansom bisected that the following commit caused the issue:
> 
> We're 100% certain of this?

Good question. I haven't yet checked it myself, it's what Kevin shared on teams.

**Comment by dsyme** [comment]
Just to note this disabled warning is being hit on unpickling

        warning(Error(FSComp.SR.pickleMissingDefinition (i, inMap.Name, ilscope.QualifiedName), range0))

which indicates something is going awry

**Comment by vzarytovskii** [comment]
> Just to note this disabled warning is being hit on unpickling
> 
>         warning(Error(FSComp.SR.pickleMissingDefinition (i, inMap.Name, ilscope.QualifiedName), range0))
> 
> which indicates something is going awry. It's being hit many times - it's not just one type definition `pos` but pretty much all intra-assembly references.
> 

Yup, both Tomáš and me noticed this today.


> The problem also seems to apply to FSharp.DependencyManager.Nuget, so it doesn't seem to indicate an assembly size problem.
> 

Huh, interesting, haven't seen it, will check it.

**Comment by dsyme** [comment]
> The problem also seems to apply to FSharp.DependencyManager.Nuget, so it doesn't seem to indicate an assembly size problem.

I'm trying to use this to get a much smaller repro, e.g. triggers early with a solution where you have nothing but these two projects

![image](https://user-images.githubusercontent.com/7204669/198404859-22137c9b-f0b5-42fd-bc1b-aacd628e08bd.png)

**Comment by dsyme** [comment]
I've made a little progress

Technique:

* One of the entities in FSharp,.DependencyManaer.Nuget hitting that assert is the first intra-assembly cross-reference to `ResolveDependenciesResult`, fsi, stamp 34588 , being written to `otycons` table slot 4

* However in the actual "definition" of that type in ModuleOrNamespaceType tables in the CCU it has stamp 34702

* Setting conditional breakpoint on NewModifiedTycon with `data.Stamp == 34702` reveals it gets copied from an earlier Entity with stamp 34628

* Setting conditional breakpoint on NewModifiedTycon with `data.Stamp == 34628` reveals it is created with this stack:

  ![image](https://user-images.githubusercontent.com/7204669/198408562-9b37487c-8a94-43f2-9ae0-402e53177b3a.png)
earlier 34628

Now, this is code that changed in that PR.  In particular that PR "simplified" the entity combining code so that it also applies to combining type definitions, though we never expected that to happen, as in the old code we would report [this error](https://github.com/dotnet/fsharp/commit/51635bbebf36e39a86d06d96730e13cfc40a095f#diff-004628b3d807a39c23e65443575bb2f125d326d99ad1852374e76788bf80b47eL10086-L10087)...... 

```
            errorR(Error(FSComp.SR.tastDuplicateTypeDefinitionInAssembly(entity2.LogicalName, textOfPath path), entity2.Range))
```

However on the stack up above we are indeed combining two type definitions both with name `ResolveDependenciesResult`! And sure enough when I set a breakpoint on that `tastDuplicateTypeDefinitionInAssembly` we are actually hitting it, and emitting that diagnostic as well!  But this is only happening in the FinalizeTypeCheckTask phase, partly because of some laziness.  So presumably we've been emitting that error condition for a while now, in the IDE< just hadn't noticed.

I'm now going to look into why we're combining two module types that both contain the same type definition.  It seems that the IDE is combining the signature and implementation for the same file, when that should never happen.

**Comment by dsyme** [comment]
Ok, it seems that the changes in that PR mean that we're calling `AddCheckResultsToTcState` twice when both `skipImplIfSigExists` is true and calling via `CheckOneInput`.  This is only ever the case via the IDE.

The double calls to AddCheckResultsToTcState are duplicating the metadata nodes for tycons improperly, causing this corrupted metadata.

I'm really surprised this hasn't been caught earlier.  It is a significant bug in the IDE.  The compiler service tests should be catching this, but it might be that the in-memory-cross-project-reference tests we have don't have rich enough code in the first project in `Project1 --> Project2`  dependencies.   The code has to be rich enough that there are intra-assembly references in the metadata of Project1, and that the elements having those cross-references are exercised in the tests being applied in Project2.

**Comment by dsyme** [comment]
I've pushed a possible fix and written some thoughts on what went wrong here in #14189

**Comment by dsyme** [comment]
@T-Gro @vzarytovskii @KevinRansom really horrible bug, thanks for the perseverance, I hope we've got it now

**Comment by vzarytovskii** [comment]
Fixed by #14189

**Comment by T-Gro** [comment]
Thanks @dsyme for a written record of your debugging approach, this does help a lot.
